### PR TITLE
fix: null value and empty string in filter

### LINKED ISSUE #18147: [explore] I can't select `NULL` and `empty string` in filter
I can't select `NULL` and `empty string` in Filter

#### How to reproduce the bug
1. First of all, let we create a virtual dataset from SQLLab and open this in Explore page.
```
SELECT '' as name
UNION ALL
SELECT null as name
```

2. I can't get `NULL` and `empty string` in filter.

image

3. I can manually input `<NULL>` and `<empty string>` to filter that.

image

### Expected results

I can select `<NULL>` and `<empty string>` in filter

### Environment

(please complete the following information):

- browser type and version:
- superset version: `superset version`
- python version: `python --version`
- node.js version: `node -v`
- any feature flags active:

### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [ ] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [ ] I have reproduced the issue with at least the latest released version of superset.
- [ ] I have checked the issue tracker for the same issue and I haven't found one similar.

### PR DESCRIPTION
**Pull request by zhaoyongjie**
### SUMMARY

closes: https://github.com/apache/superset/issues/18147
Currently, Superset Filter can't show **NULL value** and **empty string**. This PR replaced **NULL** with `<NULL>`  and replaced **""** with `<empty string>`

Notice that: A known limitation for the current design, users can't filter below text values:
```
- "<empty string>"
- "<NULL>"
```

we need to add to escape mechanism somewhere in separated PR.

### After
image

### Before
image

### TESTING INSTRUCTIONS

1. Create a virtual dataset
```
SELECT '' as name
UNION ALL
SELECT null as name
union all
SELECT 'foo' as name
```

2. Open the virtual dataset in Explore Page
3. Drag **name** column in Filter Control
4. You can get 3 options in the filter selection
5. Add all options and open query inspector. The query is:
```
SELECT count(*) AS count
FROM
  (SELECT '' as name
   UNION ALL SELECT null as name
   union all SELECT 'foo' as name) AS virtual_table
WHERE name IS NULL
  OR name IN ('',
              'foo',)
ORDER BY count DESC
LIMIT 1000;
```

### ADDITIONAL INFORMATION

- [x] Has associated issue: https://github.com/apache/superset/issues/18147
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/connectors/sqla/models.py [hunk 1]
```diff
@@ -756,9 +765,6 @@ def values_for_column(
         if limit:
             qry = qry.limit(limit)
 
-        if not contain_null:
-            qry = qry.where(target_col.get_sqla_col().isnot(None))
-
```
Comment (zhaoyongjie) [inline superset/connectors/sqla/models.py:761]: These codes were introduced by this [PR](https://github.com/apache/superset/pull/17168). After we provide **replacing value** mechanism, this snippet is unused.

**Review by villebro — COMMENTED** [review]
As a bandaid this should fix the current issue. However, I can't help but feel slightly uncomfortable by these types of workarounds. I think the optimal architectural design would be such, that `<NULL>` and `<empty string>` would be the _labels_ in the select component in the frontend, but the _value_ would still be retained in its original form. Did you consider this approach, and if so, was there some reason it doesn't work?

**Comment by zhaoyongjie** [comment]
> As a bandaid this should fix the current issue. However, I can't help but feel slightly uncomfortable by these types of workarounds. I think the optimal architectural design would be such, that `<NULL>` and `<empty string>` would be the _labels_ in the select component in the frontend, but the _value_ would still be retained in its original form. Did you consider this approach, and if so, was there some reason it doesn't work?

Nice catch! I have made frontend render of that.

File: superset/connectors/base/models.py [hunk 2]
```diff
@@ -404,7 +404,7 @@ def handle_single_value(value: Optional[FilterValue]) -> Optional[FilterValue]:
                     return utils.cast_to_num(value)
                 if value == NULL_STRING:
                     return None
-                if value == "<empty string>":
+                if value == EMPTY_STRING:
```
Comment (villebro) [inline superset/connectors/base/models.py:407]: Can we remove these now, both `EMPTY_STRING` and `NULL_STRING`? Or are these needed due in some legacy component?

File: superset/connectors/base/models.py [hunk 3]
```diff
@@ -439,9 +439,7 @@ def query(self, query_obj: QueryObjectDict) -> QueryResult:
         """
         raise NotImplementedError()
 
-    def values_for_column(
-        self, column_name: str, limit: int = 10000, contain_null: bool = True,
-    ) -> List[Any]:
+    def values_for_column(self, column_name: str, limit: int = 10000,) -> List[Any]:
```
Comment (villebro) [inline superset/connectors/base/models.py:442]: nit (I wonder if this has been fixed in a recent version of black?)
```suggestion
    def values_for_column(self, column_name: str, limit: int = 10000) -> List[Any]:
```

File: superset/connectors/druid/models.py [hunk 4]
```diff
@@ -948,9 +948,7 @@ def metrics_and_post_aggs(
         )
         return aggs, post_aggs
 
-    def values_for_column(
-        self, column_name: str, limit: int = 10000, contain_null: bool = True,
-    ) -> List[Any]:
+    def values_for_column(self, column_name: str, limit: int = 10000,) -> List[Any]:
```
Comment (villebro) [inline superset/connectors/druid/models.py:951]: same here
```suggestion
    def values_for_column(self, column_name: str, limit: int = 10000) -> List[Any]:
```

File: superset/connectors/sqla/models.py [hunk 5]
```diff
@@ -176,9 +176,7 @@ def query(self, query_obj: QueryObjectDict) -> QueryResult:
     def get_query_str(self, query_obj: QueryObjectDict) -> str:
         raise NotImplementedError()
 
-    def values_for_column(
-        self, column_name: str, limit: int = 10000, contain_null: bool = True,
-    ) -> List[Any]:
+    def values_for_column(self, column_name: str, limit: int = 10000,) -> List[Any]:
```
Comment (villebro) [inline superset/connectors/sqla/models.py:179]: and here:
```suggestion
    def values_for_column(self, column_name: str, limit: int = 10000,) -> List[Any]:
```

File: tests/integration_tests/sqla_models_tests.py [hunk 6]
```diff
@@ -477,22 +478,77 @@ def test_labels_expected_on_mutated_query(self):
...
+        with_null = table.values_for_column(column_name="foo", limit=10000)
         assert None in with_null
-        assert len(with_null) == 3
+        assert len(with_null) == 4
+
+        # null value should be replaced
+        result_object = table.query(
+            {
+                "metrics": ["count"],
+                "filter": [{"col": "foo", "val": [NULL_STRING], "op": "IN"}],
```
Comment (villebro) [inline tests/integration_tests/sqla_models_tests.py:501]: Should we not test that we can query a real `None` value here?

File: tests/integration_tests/sqla_models_tests.py [hunk 7]
```diff
@@ -477,22 +478,77 @@ def test_labels_expected_on_mutated_query(self):
...
+                "is_timeseries": False,
+            }
+        )
+        assert result_object.df["count"][0] == 1
+
+        # empty string should be replaced
+        result_object = table.query(
+            {
+                "metrics": ["count"],
+                "filter": [{"col": "foo", "val": [EMPTY_STRING], "op": "IN"}],
```
Comment (villebro) [inline tests/integration_tests/sqla_models_tests.py:521]: ..and `""` here?

**Review by villebro — COMMENTED** [review]
This is great! One last question regarding possible simplification of `handle_single_value`, other than that LGTM

File: superset-frontend/src/explore/components/controls/FilterControl/AdhocFilterEditPopoverSimpleTabContent/index.tsx [hunk 8]
```diff
@@ -346,7 +349,12 @@ const AdhocFilterEditPopoverSimpleTabContent: React.FC<Props> = props => {
           endpoint: `/superset/filter/${datasource.type}/${datasource.id}/${col}/`,
         })
           .then(({ json }) => {
-            setSuggestions(json);
+            setSuggestions(
+              json.map((suggestion: null | number | boolean | string) => ({
+                value: suggestion,
+                label: optionLabel(suggestion),
```
Comment (michael-s-molina) [inline superset-frontend/src/explore/components/controls/FilterControl/AdhocFilterEditPopoverSimpleTabContent/index.tsx:355]: Shouldn't we translate the option labels? I notice that the `optionLabel` function is always returning the values in English.

Comment (villebro) [reply superset-frontend/src/explore/components/controls/FilterControl/AdhocFilterEditPopoverSimpleTabContent/index.tsx:355]: ..then the translation needs to be done where the constant is defined. And if we do that, then we should definitely remove support for using the strings `"<NULL>"` and `"<empty string>"` as proxies for `None` and `""` respectively in the backend and assume they'll always just be labels, but never values.

File: superset/connectors/base/models.py (hunk 2 above)
Comment (zhaoyongjie) [reply superset/connectors/base/models.py:407]: I think it's used to adapt some legacy Select component.

File: tests/integration_tests/sqla_models_tests.py (hunk 7 above)
Comment (zhaoyongjie) [reply tests/integration_tests/sqla_models_tests.py:521]: "" and EMPTY_STRING both in this test case

File: superset/connectors/base/models.py (hunk 3 above)
Comment (zhaoyongjie) [reply superset/connectors/base/models.py:442]: It seems not. Manually fixed

File: tests/integration_tests/sqla_models_tests.py (hunk 6 above)
Comment (zhaoyongjie) [reply tests/integration_tests/sqla_models_tests.py:501]: NULL and NULL_STRING both in this test case

**Review by villebro — APPROVED** [review]
LGTM - there's some additional ideas on how this could be further improved, but let's leave those to follow-up PRs to get this bug fixed first. Thanks @zhaoyongjie for all the iteration, much appreciated!

**Comment by zhaoyongjie** [comment]
> LGTM - there's some additional ideas on how this could be further improved, but let's leave those to follow-up PRs to get this bug fixed first. Thanks @zhaoyongjie for all the iteration, much appreciated!

there's a related [PR](https://github.com/apache/superset/pull/17881), I will work together with @stephenLYZ  to improve this.

**Pull request merged**